lib/ansible/plugins/filter/core.py

```diff
@@ -326,7 +326,7 @@ def combine(*terms, **kwargs):
 
     dicts = []
     for t in terms:
-        if isinstance(t, MutableMapping):
+        if isinstance(t, (MutableMapping, Mapping)):
```

Comment:
If this is needed, just replace MutableMapping with Mapping.  (MutableMapping is a subclass of Mapping that adds additional features).

OTOH, I wonder why we're dealing with something that's just a Mapping instead of a Mutable Mapping in the first place....

Reply:
its due to the hostvars object, which inherits from mapping .. i doubt we want it to be mutable.

Reply:
If we allow non-Mutable mappings here, then we need to put a try: except around the call to reduce in the if recurse: later as giving it two non-mutable mappings will throw an error.

I dislike that non-mutable mappings would be accepted with recurse=False but cause an error with recurse=True so perhaps we should track down why we're getting non-mutable mappings or perhaps we should enhance merge_hash to be able to operate on non-mutable mappings.

Reply:
<nod>  Well.  using this function with hostvars and recursive=True will traceback if this is all we do.  So there needs to be more work done.  I think that the right way to fix it probably means changing merge_hash to copy Mapping a into a MutableMapping instead of calling copy().  I'm not sure how much that will slow down critical code, though....

So maybe what we should do is handle the non-mutable case in this function.  Maybe something like:

``` python

def combine(*terms, **kwargs):
    recursive = kwargs.get('recursive', False)
    if len(kwargs) > 1 or (len(kwargs) == 1 and 'recursive' not in kwargs):
        raise AnsibleFilterError("'recursive' is the only valid keyword argument")

    dicts = []
    for t in terms:
        if isinstance(t, Mapping):
            if recursive and not isinstance(t, MutableMapping):
                t = dict(t)
            dicts.append(t)
        elif isinstance(t, list):
            dicts.append(combine(*t, **kwargs))
        else:
            raise AnsibleFilterError("|combine expects dictionaries, got " + repr(t))

    if recursive:
        return reduce(merge_hash, dicts)
    else:
        return dict(itertools.chain(*map(iteritems, dicts)))
```

Reply:
The use-case that this fixes is already an odd-ball one. Throwing `recursive=True` into the mix as well probably *should* break.

Another option is just to make `HostVarsVars` a `MutableMapping` (if it's even necessary) but have its `__setitem__()` function throw an exception saying that it's *supposed* to be read-only. Using `|combine()` on `hostvars[host]` (or writing to vars via `hostvars` at all) is probably not something that we want to support.

Reply:
I don't think so... if we support doing this on hostvars at all, then there's no obvious reason to the user that doing it with recursive=True should fail.

Making HostVars a MutableMapping but making __setitem__ throw an exception will still break with recursive=True.

So I think the code I posted is probably the best way to do it as it obeys the semantics of the combine filter without slowing down the common case for merge_hash().  If it's okay to slow down merge_hash(), then we can implement the promotion from non-mutable to mutable inside of that function but I kinda doubt that's the case since we've survived this long without it.